src/scripts/emailUpperCaseMigration.js

```diff
@@ -0,0 +1,146 @@
+        const lowerCasedSha1 = await subscribeLowercaseHashToHIBP(
+          subRecord.primary_email,
+        );
+        await knex("subscribers")
+          .update({
+            primary_sha1: lowerCasedSha1,
+          })
+          .where("id", subRecord.id)
+          .transacting(trx);
+
+        // update primary email to lowercase
+        await knex("subscribers")
+          .update({
+            primary_email: subRecord.primary_email.toLowerCase(),
+          })
+          .where("id", subRecord.id)
+          .transacting(trx);
+        await trx.commit();
+        console.log(`fixed subscribers record ID: ${subRecord.id}`);
+      } catch (e) {
+        await trx.rollback();
+        console.error(`Error fixing record: ${subRecord.id}`, e);
+      }
+
+      subPrevMaxId = subRecord.id;
+      subRecordsFixed++;
+      console.log(
+        `subscribed lower-case address hash for subscribers record ID: ${subRecord.id}`,
+      );
+    }
+  }
+
+  /**
+   * Fixing Emails
+   */
+  console.log(`/** Emails Table **/`);
+
+  const emailRecordsThatNeedFixing = await knex("email_addresses")
+    .count()
+    .whereRaw("email != lower(email)");
+  const emailWithUpperCount = emailRecordsThatNeedFixing[0].count;
+  console.log(
+    `found ${emailWithUpperCount} email_address records with email != lower(email). fixing ...`,
+  );
+
+  let emailRecordsFixed = 0;
+  let emailPrevMaxId = 0;
+  while (emailRecordsFixed < emailWithUpperCount) {
+    console.log(`working on chunk where id > ${emailPrevMaxId} ...`);
+    const emailRecordsWithUpperChars = await knex
+      .select("id", "email")
+      .from("email_addresses")
+      .where("id", ">", emailPrevMaxId)
+      .whereRaw("email != lower(email)")
+      .orderBy("id", "asc")
+      .limit(chunkSize);
+    for (const emailRecord of emailRecordsWithUpperChars) {
+      console.log(`Fixing record: ${emailRecord.id}`);
+      const trx = await knex.transaction();
+      try {
+        // update sha1 and sub to HIBP with the new sha1
+        const lowerCasedSha1 = await subscribeLowercaseHashToHIBP(
+          emailRecord.email,
+        );
+        await knex("email_addresses")
+          .update({
+            sha1: lowerCasedSha1,
+          })
+          .where("id", emailRecord.id)
+          .transacting(trx);
+
+        // update primary email to lowercase
+        await knex("email_addresses")
+          .update({
+            email: emailRecord.email.toLowerCase(),
+          })
+          .where("id", emailRecord.id)
+          .transacting(trx);
+        await trx.commit();
+        console.log(`fixed email_addresses record ID: ${emailRecord.id}`);
+      } catch (e) {
+        await trx.rollback();
+        console.error(
+          `Error fixing email_addresses record: ${emailRecord.id}`,
+          e,
+        );
+      }
+
+      emailPrevMaxId = emailRecord.id;
+      emailRecordsFixed++;
+    }
+  }
+
+  console.log("done.");
+  process.exit();
```

Comment:
Should this script exit with a non-zero error code if we hit a `catch` block and rollback a transaction?

Reply:
Since this will be a one-off script, we will be watching the output and rerun if we run into any rollbacks.